File: internal/teaweb/actions/default/settings/mongo/update.go
package mongo

import (
	"github.com/TeaWeb/build/internal/teaconfigs/db"
	"github.com/TeaWeb/build/internal/teadb"
	"github.com/iwind/TeaGo/actions"
	"github.com/iwind/TeaGo/maps"
)

type UpdateAction actions.Action

// 修改连接
func (this *UpdateAction) Run(params struct {
	Action string
}) {
	this.Data["action"] = params.Action
	this.Data["typeIsChanged"] = db.SharedDBConfig().Type != db.DBTypeMongo

	config, err := db.LoadMongoConfig()
	if err != nil {
		config = db.DefaultMongoConfig()
	}
	this.Data["config"] = maps.Map{
		"scheme":                  config.Scheme,
		"username":                config.Username,
		"password":                "",
		"host":                    config.Host(),
		"port":                    config.Port(),
		"dbName":                  config.DBName,
		"authEnabled":             config.AuthEnabled,
		"authMechanism":           config.AuthMechanism,
		"authMechanismProperties": config.AuthMechanismPropertiesString(),
		"requestURI":              config.RequestURI,
		"poolSize":                config.PoolSize,
		"timeout":                 config.Timeout,
	}

	this.Show()
}

// 提交保存
func (this *UpdateAction) RunPost(params struct {
	Host                    string
	Port                    uint
	DBName                  string `alias:"dbName"`
	Username                string
	Password                string
	AuthEnabled             bool
	AuthMechanism           string
	AuthMechanismProperties string
	PoolSize                int
	Timeout                 int

	Must *actions.Must
}) {
	// 是否已改变
	sharedConfig := db.SharedDBConfig()
	isChanged := sharedConfig.Type != db.DBTypeMongo

	params.Must.
		Field("host", params.Host).
		Require("请输入主机地址").
		Field("port", params.Port).
		Require("请输入端口").
		Gt(0, "请输入正确的端口").
		Field("poolSize", params.PoolSize).
		Gte(0, "连接池不能小于0").
		Field("timeout", params.Timeout).
		Gte(0, "超时时间不能小于0")

	config, err := db.LoadMongoConfig()
	if err != nil {
		this.Fail(err.Error())
	}

	config.SetAddr(params.Host, params.Port)
	config.DBName = params.DBName
	config.AuthEnabled = params.AuthEnabled
	config.AuthMechanism = params.AuthMechanism
	config.LoadAuthMechanismProperties(params.AuthMechanismProperties)
	config.Username = params.Username
	if len(params.Username) > 0 {
		if len(params.Password) > 0 {
			config.Password = params.Password
		}
	} else {
		config.Password = ""
	}
	config.PoolSize = params.PoolSize
	config.Timeout = params.Timeout
	err = config.Save()

	if err != nil {
		this.Fail("保存失败：" + err.Error())
	}

	if isChanged {
		shouldRestart = true

		sharedConfig.Type = db.DBTypeMongo
		err = sharedConfig.Save()
		if err != nil {
			this.Fail("保存失败：" + err.Error())
		}
	}

	// 重新连接
	teadb.ChangeDB()

	this.Next("/settings/mongo", nil).Success("保存成功")
}

File: web/views/@default/settings/mongo/update.js
Tea.context(function () {
	this.isTesting = false;
	this.testingError = "";
	this.testingSuccess = "";

	if (this.config.authMechanism.length == 0) {
		this.config.authMechanism = "SCRAM-SHA-1";
	}

	this.updateAuth = function () {
		this.config.authEnabled = !this.config.authEnabled;
	};

	this.testConnection = function () {
		var params = {
			host: this.config.host,
			port: this.config.port,
			dbName: this.config.dbName,
			authEnabled: this.config.authEnabled ? 1 : 0
		};
		if (this.config.authEnabled) {
			if (this.config.username.length == 0) {
				alert("请输入用户名");
				this.$find("form input[name='username']").focus();
				return;
			}
			params["username"] = this.config.username;
			params["password"] = this.config.password;
			params["authMechanism"] = this.config.authMechanism;
			params["authMechanismProperties"] = this.config.authMechanismProperties;
		}

		this.isTesting = true;
		this.testingError = "";
		this.testingSuccess = "";

		this.$get(".test")
			.params(params)
			.timeout(10)
			.success(function () {
				this.testingError = "";
				this.testingSuccess = "连接成功！";
			})
			.fail(function (resp) {
				if (resp) {
					this.testingError = resp.message;
				}
			})
			.error(function () {
				this.testingError = "连接超时";
			})
			.done(function () {
				this.isTesting = false;
			});
	};

	/**
	 * 高级选项
	 */
	this.advancedOptionsVisible = false;

	this.showAdvancedOptions = function () {
		this.advancedOptionsVisible = !this.advancedOptionsVisible;
	};
});